File: src/Utils/Color.cpp

#include "Color.hpp"

namespace Ilya
{

    const Color Color::Black {0.f, 0.f, 0.f};
    const Color Color::White {1.f, 1.f, 1.f};
    const Color Color::Red {1.f, 0.f, 0.f};
    const Color Color::Green {0.f, 1.f, 0.f};
    const Color Color::Blue {0.f, 0.f, 1.f};
    const Color Color::Yellow {0.f, 1.f, 1.f};
    const Color Color::Magenta {1.f, 0.f, 1.f};
    const Color Color::Brown {1.f, 1.f, 0.f};

    Color::Color(float factor): r(factor), g(factor), b(factor), a(1.f)
    {}

    Color::Color(float r, float g, float b, float a): r(r), g(g),
                                                      b(b), a(a)
    {}

    Color::Color(const Vector3& color): r(color.x), g(color.y),
                                        b(color.z), a(1.f)
    {}

    Color::Color(const Vector4& color): r(color.x), g(color.y),
                                        b(color.z), a(color.w)
    {}

    Color& Color::operator+=(const Color& color)
    {
        r += color.r;
        g += color.g;
        b += color.b;
        a += color.a;

        return *this;
    }

    Color& Color::operator-=(const Color& color)
    {
        r -= color.r;
        g -= color.g;
        b -= color.b;
        a -= color.a;

        return *this;
    }

    Color& Color::operator*=(const Color& color)
    {
        r *= color.r;
        g *= color.g;
        b *= color.b;
        a *= color.a;

        return *this;
    }

    Color& Color::operator*=(float factor)
    {
        r *= factor;
        g *= factor;
        b *= factor;
        a *= factor;

        return *this;
    }

    Color::operator Vector3() const
    {
        return Vector3(r, g, b);
    }

    Color::operator Vector4() const
    {
        return Vector4(r, g, b, a);
    }

    Color operator+(const Color& c1, const Color& c2)
    {
        return { c1.r + c2.r, c1.g + c2.g, c1.b + c2.b, c1.a + c2.a };
    }

    Color operator-(const Color& c1, const Color& c2)
    {
        return { c1.r - c2.r, c1.g - c2.g, c1.b - c2.b, c1.a - c2.a };
    }

    Color operator*(const Color& c1, const Color& c2)
    {
        return { c1.r * c2.r, c1.g * c2.g, c1.b * c2.b, c1.a * c2.a };
    }

    Color operator*(const Color& color, float factor)
    {
        return { color.r * factor, color.g * factor, color.b * factor, color.a };
    }

    Color operator*(float factor, const Color& color)
    {
        return { color.r * factor, color.g * factor, color.b * factor, color.a };
    }
}

File: src/color.rs
use std::ops;

pub struct Color {
    pub r: f32,
    pub g: f32,
    pub b: f32,
    pub a: f32,
}

impl Color {
    pub const WHITE: Self = Self::rgb(1.0, 1.0, 1.0);
    pub const RED: Self = Self::rgb(1.0, 0.0, 0.0);
    pub const GREEN: Self = Self::rgb(0.0, 1.0, 0.0);
    pub const BLUE: Self = Self::rgb(0.0, 0.0, 1.0);
    pub const BLACK: Self = Self::rgb(0.0, 0.0, 0.0);

    pub const fn rgb(r: f32, g: f32, b: f32) -> Self {
        Self { r, g, b, a: 1.0 }
    }

    pub const fn rgba(r: f32, g: f32, b: f32, a: f32) -> Self {
        Self { r, g, b, a }
    }

    pub fn to_255(&self) -> [u8; 4] {
        let r = f32::clamp(self.r * 255.99, 0.0, 255.99) as u8;
        let g = f32::clamp(self.g * 255.99, 0.0, 255.99) as u8;
        let b = f32::clamp(self.b * 255.99, 0.0, 255.99) as u8;
        let a = f32::clamp(self.a * 255.99, 0.0, 255.99) as u8;

        [r, g, b, a]
    }
}

impl ops::Mul<f32> for Color {
    type Output = Self;

    fn mul(self, rhs: f32) -> Self {
        Self::rgb(
            self.r * rhs,
            self.g * rhs,
            self.b * rhs,
        )
    }
}

impl ops::Mul<Color> for f32 {
    type Output = Color;

    fn mul(self, rhs: Color) -> Color {
        Color::rgb(
            self * rhs.r,
            self * rhs.g,
            self * rhs.b,
        )
    }
}

impl ops::Add<Color> for Color {
    type Output = Self;

    fn add(self, rhs: Color) -> Self {
        Self::rgb(
            self.r + rhs.r,
            self.g + rhs.g,
            self.b + rhs.b,
        )
    }
}